orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+
+        trainer_options = 'trainer_options'
+        states[trainer_options] = {}
+        states[trainer_options]['mixed_precision'] = self.options.mixed_precision.enabled
+        states[trainer_options]['zero_stage'] = self.options.distributed.deepspeed_zero_optimization.stage or 0
+        states[trainer_options]['world_rank'] = self.options.distributed.world_rank or 0
+        states[trainer_options]['world_size'] = self.options.distributed.world_size or 1
+        states[trainer_options]['optimizer_name'] = self.optim_config.name
+
+    def state_dict(self, pytorch_format=False):
+        """Extracts model and optimizer states from the training session and returns them in a dictionary
+
+        - Extracts model and optimizer states from the training session
+        - Extracts relevant trainer options
+        - Extracts partition information from the training session in case of a ZeRO enabled distributed
+            ORTTrainer configuration
+        - Loads them into a dictionary and returns this dictionary
+
+        Args:
+            pytorch_format: boolean flag to indicate the format of the state dictionary to return.
+
+        Returns:
+            a dictionary with loaded states
+            structure of the returned dictionary:
+            - when pytorch_format is false
+                {
+                    "model":
+                    {
+                        "fp32":
+                        {
+                            model_weight_name:  model_weight_tensor
+                        }
+                    },
+                    "optimizer":
+                    {
+                        model_weight_name:
+                        {
+                            "Moment_1": moment1_tensor,
+                            "Moment_2": moment2_tensor,
+                            "Update_Count": update_tensor # if optimizer is adam, absent otherwise
+                        },
+                        "shared_optimizer_state": # if optimizer is shared, absent otherwise.
+                        {
+                            "step": step_tensor # int array of size 1
+                        }
+                    }
+                    "trainer_options":
+                    {
+                        "mixed_precision": mixed_precision_enabled_bool,
+                        "zero_stage": zero_enabled_stage,
+                        "world_rank": world_rank_int,
+                        "world_size": world_size_int,
+                        "optimizer_name": training_optimizer_name
+                    },
+                    "partition_info":
+                    {
+                        model_weight_name:
+                        {
+                            "original_dim": original_dimension_array
+                        }
+                    }
+                }
+            - when pytorch_format is true
+                {
+                    model_weight_name: model_weight_tensor
+                }
```

Comment:
Markdown doesn't work with Returns with multiple lines, so better moving this schema structure to the main docstring body, before Args:  section